Compiler/NFFrontEnd/NFTypeCheck.mo

```diff
@@ -621,6 +622,229 @@ public function checkBinaryOperationArrays
+          arrtp := promoteLeft(inType2,dims);
+
+          (exp1,exp2,ty1) := matchTypeBothWays(inExp1,inType1,inExp2,arrtp);
+
+          // Create a scalar Real Type and lift it to array.
+          // Necessary because even if both operands are of Integer type the result
+          // should be Real type with dimensions of the input array operand.
+          typsrc := Types.getTypeSource(ty1);
+          ty := DAE.T_REAL({},typsrc);
+
+          outType := promoteLeft(ty,dims);
+          outExp := DAE.BINARY(exp1, DAE.DIV_ARRAY_SCALAR(outType), exp2);
+        else
+          binaryArrayOpError(inExp1,inType1,inExp2,inType2,inOp,
+            "\n: Dividing a sclar by array or array by array is not a valid " +
+            "operation in Modelica. Try using elementwise operator './'");
+          fail();
+        end if;
+      then
+        (outExp, outType);
+
+    case DAE.POW()
+      algorithm
+        if (nrDims1 == 2 and not isArray2) then
+          M := Types.getDimensionNth(inType1, 1);
+          K := Types.getDimensionNth(inType1, 2);
+          // Check if dims are equal. i.e Square Matrix
+          if not(isValidMatrixMultiplyDims(M, K)) then
+            binaryArrayOpError(inExp1,inType1,inExp2,inType2,inOp,
+              "\n: Exponentiation involving arrays is only valid for square " +
+              "matrices with integer exponents. Try using elementwise operator '.^'");
+            fail();
+          end if;
+
+          try
+            DAE.T_INTEGER(_,_) := inType2;
+          else
+            binaryArrayOpError(inExp1,inType1,inExp2,inType2,inOp,
+              "\n: Exponentiation involving arrays is only valid for square " +
+              "matrices with integer exponents. Try using elementwise operator '.^'");
+            fail();
+          end try;
+
+          outType := inType1;
+          outExp := DAE.BINARY(inExp1, DAE.POW_ARRAY_SCALAR(inType1), inExp2);
+        else
+          binaryArrayOpError(inExp1,inType1,inExp2,inType2,inOp,
+            "\n: Exponentiation involving arrays is only valid for square " +
+            "matrices with integer exponents. Try using elementwise operator '.^'");
+          fail();
+        end if;
+      then
+        (outExp, outType);
+
+
+    case DAE.MUL()
+      algorithm
+        if (not isArray1 or not isArray2) then
+
+          arrtp := if isArray1 then inType1 else inType2;
+          dims := getArrayTypeDims(arrtp);
+          //match their scalar types. For now.
+          ty1 := underlyingType(inType1);
+          ty2 := underlyingType(inType2);
+          // TODO: one of the exps is array but its type is now simple.
+          (exp1,exp2,ty) := matchTypeBothWays(inExp1,ty1,inExp2,ty2);
+
+          outType := promoteLeft(ty,dims);
+          outExp := DAE.BINARY(exp1, DAE.MUL_ARRAY_SCALAR(outType), exp2);
+
+        elseif (nrDims1 == 1 and nrDims2 == 1) then
+          {N1} := getArrayTypeDims(inType1);
+          {N2} := getArrayTypeDims(inType1);
```

Comment:
Ya this should be inType2. Probably a couple more of errors like this exist. Will filter them out when we have tests in place. Will fix this for now.